File: customer-java-resttemplate/src/test/java/com/prez/UsingMongoDBAndRedis.java
package com.prez;

import org.springframework.boot.test.util.TestPropertyValues;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.test.context.ContextConfiguration;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.utility.DockerImageName;

@ContextConfiguration(initializers = {UsingMongoDBAndRedis.Initializer.class})
public class UsingMongoDBAndRedis {
    public static final int MONGODB_PORT = 27017;
    public static final int REDIS_PORT = 6379;

    @Container
    public static GenericContainer<?> mongodb = new GenericContainer<>(DockerImageName.parse("mongo:4.2"))
            .withExposedPorts(MONGODB_PORT); // we don't reuse container for integration tests because we get timeouts

    @Container
    public static GenericContainer<?> redis = new GenericContainer<>(DockerImageName.parse("redis:6.2"))
            .withExposedPorts(REDIS_PORT); // we don't reuse container for integration tests because we get timeouts

    static class Initializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {
        @Override
        public void initialize(ConfigurableApplicationContext applicationContext) {
            redis.start();
            mongodb.start();

            TestPropertyValues.of(
                    "spring.data.mongodb.uri=mongodb://127.0.0.1:" + mongodb.getMappedPort(MONGODB_PORT),
                    "spring.data.mongodb.database=customerdbtest",
                    "spring.redis.port=" + redis.getMappedPort(REDIS_PORT)
            ).applyTo(applicationContext.getEnvironment());
        }
    }

}

File: customer-kotlin-webflux/src/test/kotlin/com/prez/UsingMongoDBAndRedis.kt
package com.prez

import org.springframework.boot.test.util.TestPropertyValues
import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.test.context.ContextConfiguration
import org.testcontainers.containers.GenericContainer
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.utility.DockerImageName

@ContextConfiguration(initializers = [UsingMongoDBAndRedis.Initializer::class])
open class UsingMongoDBAndRedis {

  companion object {
    const val MONGODB_PORT = 27017
    const val REDIS_PORT = 6379

    @Container
    var mongodb: GenericContainer<*> = GenericContainer<Nothing>(DockerImageName.parse("mongo:4.0"))
        .withExposedPorts(MONGODB_PORT)

    @Container
    var redis: GenericContainer<*> = GenericContainer<Nothing>(DockerImageName.parse("redis:6.0.9"))
        .withExposedPorts(UsingRedis.REDIS_PORT)
  }

  internal class Initializer : ApplicationContextInitializer<ConfigurableApplicationContext> {
    override fun initialize(applicationContext: ConfigurableApplicationContext) {
      // redis.withReuse(true) // we don't reuse container for integration tests because we get timeouts
      redis.start()
      // mongodb.withReuse(true)
      mongodb.start()
      TestPropertyValues.of(
          "spring.data.mongodb.uri=mongodb://127.0.0.1:" + mongodb.getMappedPort(MONGODB_PORT),
          "spring.data.mongodb.database=customerdbtest",
          "spring.redis.port=" + redis.getMappedPort(REDIS_PORT)
      ).applyTo(applicationContext.environment)
    }
  }

}
